File: Maintenance.ascx.vb
'
' DotNetNuke® - http://www.dotnetnuke.com
' Copyright (c) 2002-2011 by DotNetNuke Corp. 

'
' Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
' documentation files (the "Software"), to deal in the Software without restriction, including without limitation 
' the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and 
' to permit persons to whom the Software is furnished to do so, subject to the following conditions:
'
' The above copyright notice and this permission notice shall be included in all copies or substantial portions 
' of the Software.
'
' THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED 
' TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL 
' THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
' CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
' DEALINGS IN THE SOFTWARE.
'

Option Strict On

Imports System.IO
Imports DotNetNuke
Imports DotNetNuke.Security
Imports DotNetNuke.Entities.Portals
Imports DotNetNuke.Entities.Modules
Imports DotNetNuke.Entities.Modules.Actions
Imports DotNetNuke.Common.Globals
Imports DotNetNuke.Entities.Users
Imports DotNetNuke.Services.Localization
Imports DotNetNuke.Modules.Gallery.Utils
Imports DotNetNuke.Modules.Gallery.Config
Imports DotNetNuke.Modules.Gallery.WebControls

Namespace DotNetNuke.Modules.Gallery

  Public MustInherit Class Maintenance
    Inherits DotNetNuke.Entities.Modules.PortalModuleBase
    Implements DotNetNuke.Entities.Modules.IActionable
    Public lnkViewText As String = ""

#Region "Private Members"
    ' Obtain PortalSettings from Current Context   
    Dim _portalSettings As PortalSettings = PortalController.GetCurrentPortalSettings
    Private mRequest As GalleryMaintenanceRequest
    Private mGalleryConfig As DotNetNuke.Modules.Gallery.Config
    Private mFolder As GalleryFolder
    Private mGalleryAuthorization As Gallery.Authorization
    Private mSelectAll As Boolean
    Private mReturnCtl As String

    'For Localization of FileInfo text
    Private mLocalizedAlbumText As String
    Private mLocalizedSourceText As String
    Private mLocalizedThumbText As String
    Private mLocalizedMissingText As String
    Private mLocalizedPresentText As String
    Private mLocalizedRebuildThumbText As String
    Private mLocalizedCopySourceToFileText As String
    Private mLocalizedCopyFileToSourceText As String
    Private mYellowWarningImageUrl As String
    Private mGreenOKImageUrl As String

#End Region

#Region " Web Form Designer Generated Code "

    'This call is required by the Web Form Designer.
    <System.Diagnostics.DebuggerStepThrough()> Private Sub InitializeComponent()

    End Sub

    Private Sub Page_Init(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Init
      'CODEGEN: This method call is required by the Web Form Designer
      'Do not modify it using the code editor.
      InitializeComponent()

      If mGalleryConfig Is Nothing Then
        mGalleryConfig = GetGalleryConfig(ModuleId)
      End If
      'mImageFolder = "../" & mGalleryConfig.ImageFolder.Substring(mGalleryConfig.ImageFolder.IndexOf("/DesktopModules"))

    End Sub

#End Region

#Region "Optional Interfaces"

    Public ReadOnly Property ModuleActions() As ModuleActionCollection Implements Entities.Modules.IActionable.ModuleActions
      Get
        Dim Actions As New ModuleActionCollection
        Actions.Add(GetNextActionID, Localization.GetString("Configuration.Action", LocalResourceFile), ModuleActionType.ModuleSettings, "", "edit.gif", EditUrl(ControlKey:="configuration"), "", False, SecurityAccessLevel.Admin, True, False)
        Actions.Add(GetNextActionID, Localization.GetString("GalleryHome.Action", LocalResourceFile), ModuleActionType.ContentOptions, "", "icon_moduledefinitions_16px.gif", NavigateURL(), False, SecurityAccessLevel.Edit, True, False)
        Return Actions
      End Get
    End Property

#End Region

    Public ReadOnly Property ReturnCtl() As String
      Get
        Return "Maintenance;"
      End Get
    End Property

    Public ReadOnly Property GalleryConfig() As DotNetNuke.Modules.Gallery.Config
      Get
        Return mGalleryConfig
      End Get
    End Property

    Private Sub Page_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

      mRequest = New GalleryMaintenanceRequest(ModuleId)
      mGalleryConfig = mRequest.GalleryConfig
      mFolder = mRequest.Folder

      mGalleryAuthorization = New Authorization(ModuleId)
      If Not mGalleryAuthorization.HasItemEditPermission(mFolder) Then
        Response.Redirect(AccessDeniedURL(Localization.GetString("Insufficient_Maintenance_Permissions", mGalleryConfig.SharedResourceFile)))
      End If

      ' Load the styles
      DirectCast(Page, CDefault).AddStyleSheet(CreateValidID(GalleryConfig.Css()), GalleryConfig.Css())

      lnkViewText = Localization.GetString("lnkView.Text", LocalResourceFile)
      mLocalizedAlbumText = Localization.GetString("Album.Header", LocalResourceFile)
      mLocalizedSourceText = Localization.GetString("Source.Header", LocalResourceFile)
      mLocalizedThumbText = Localization.GetString("Thumb.Header", LocalResourceFile)
      mLocalizedMissingText = Localization.GetString("Missing", LocalResourceFile)
      mLocalizedPresentText = Localization.GetString("Present", LocalResourceFile)
      mLocalizedRebuildThumbText = Localization.GetString("RebuildThumb", LocalResourceFile)

      mYellowWarningImageUrl = "~/DesktopModules/Gallery/Images/yellow-warning.gif"
      mGreenOKImageUrl = "~/DesktopModules/Gallery/Images/green-ok.gif"

      'William Severance - Added condition Not Page.IsPostBack to fix loss of
      'datagrid header localization following postback. LocalizeDataGrid must be called
      'before datagrid is data bound.

      If Not Page.IsPostBack Then Localization.LocalizeDataGrid(grdContent, LocalResourceFile)

      ' Load gallery menu, this method allow call request only once
      Dim galleryMenu As GalleryMenu = CType(LoadControl("Controls/ControlGalleryMenu.ascx"), GalleryMenu)
      With galleryMenu
        .ModuleID = ModuleId
        .GalleryRequest = CType(mRequest, BaseRequest)
      End With
      celGalleryMenu.Controls.Add(galleryMenu)

      ' Load gallery breadcrumbs, this method allow call request only once
      Dim galleryBreadCrumbs As BreadCrumbs = CType(LoadControl("Controls/ControlBreadCrumbs.ascx"), BreadCrumbs)
      With galleryBreadCrumbs
        .ModuleID = ModuleId
        .GalleryRequest = CType(mRequest, BaseRequest)
      End With
      celBreadcrumbs.Controls.Add(galleryBreadCrumbs)

      If Not Page.IsPostBack AndAlso mGalleryConfig.IsValidPath Then
        If Not mFolder.IsPopulated Then
          Response.Redirect(ApplicationURL)
        End If

        BindData()

        btnDeleteAll.Attributes.Add("onClick", "javascript: return confirm('Are you sure you wish to delete selected items?')")

      End If

    End Sub

    Private Sub BindData()

      With mFolder
        txtPath.Text = mFolder.URL
        txtName.Text = .Name
      End With

      lblAlbumInfo.Text = AlbumInfo()
      BindChildItems()

    End Sub

    Protected Function BrowserURL(ByVal DataItem As Object) As String
      Dim item As GalleryMaintenanceFile = CType(DataItem, GalleryMaintenanceFile)
      If mGalleryConfig.AllowPopup Then
        Return item.URL
      Else
        Dim params As New Generic.List(Of String)
        params.Add("returnctl=" & ReturnCtl)
        If Not Request.QueryString("currentstrip") Is Nothing Then params.Add("currentstrip=" & Request.QueryString("currentstrip"))
        Return Utils.AppendURLParameters(item.URL, params.ToArray())
      End If
    End Function

    Protected Function FileInfo(ByVal DataItem As Object) As String
      Dim file As GalleryMaintenanceFile = CType(DataItem, GalleryMaintenanceFile)
      Dim sb As New System.Text.StringBuilder

      If Not file.SourceExists Then sb.Append(mLocalizedSourceText)

      If Not file.ThumbExists Then
        If sb.Length > 0 Then sb.Append(", ")
        sb.Append(mLocalizedThumbText)
      End If

      If Not file.FileExists Then
        If sb.Length > 0 Then sb.Append(", ")
        sb.Append(mLocalizedAlbumText)
      End If

      If sb.Length > 0 Then
        sb.Append(" ")
        sb.Append(mLocalizedMissingText)
      End If

      Return sb.ToString
    End Function

    Protected Function StatusImage(ByVal DataItem As Object, ByVal FileType As String) As String
      Dim file As GalleryMaintenanceFile = CType(DataItem, GalleryMaintenanceFile)
      Dim exists As Boolean = False
      Select Case FileType
        Case "Thumb"
          exists = file.ThumbExists
        Case "Source"
          exists = file.SourceExists
        Case "File"
          exists = file.FileExists
        Case Else
          Throw New ArgumentException("Must be 'Thumb', Source' or 'File'", "FileType")
      End Select
      If exists Then
        Return mGreenOKImageUrl
      Else
        Return mYellowWarningImageUrl
      End If
    End Function

    Protected Function Tooltip(ByVal DataItem As Object, ByVal FileType As String) As String
      Dim file As GalleryMaintenanceFile = CType(DataItem, GalleryMaintenanceFile)
      Dim localizedTooltip As String = mLocalizedPresentText

      Select Case FileType
        Case "Thumb"
          If Not file.ThumbExists Then localizedTooltip = mLocalizedRebuildThumbText
        Case "Source"
          If Not file.SourceExists Then localizedTooltip = mLocalizedCopyFileToSourceText
        Case "File"
          If Not file.FileExists Then localizedTooltip = mLocalizedCopySourceToFileText
        Case Else
          Throw New ArgumentException("Must be 'Thumb', Source' or 'File'", "FileType")
      End Select
      Return localizedTooltip
    End Function

    Protected Function AlbumInfo() As String
      Dim itemCount As Integer = mRequest.ImageList.Count
      Dim sourceCount As Integer = 0
      Dim albumCount As Integer = 0
      Dim thumbCount As Integer = 0
      Dim sb As New System.Text.StringBuilder

      For Each file As GalleryMaintenanceFile In mRequest.ImageList
        If file.SourceExists Then sourceCount += 1
        If file.FileExists Then albumCount += 1
        If file.ThumbExists Then thumbCount += 1
      Next

      sb.Append(Localization.GetString("AlbumInfo", LocalResourceFile))
      sb.Replace("[ItemCount]", itemCount.ToString)
      sb.Replace("[AlbumName]", mRequest.Folder.Name)
      sb.Replace("[ItemNoSource]", (itemCount - sourceCount).ToString)
      sb.Replace("[ItemNoAlbum]", (itemCount - albumCount).ToString)
      sb.Replace("[ItemNoThumb]", (itemCount - thumbCount).ToString)
      sb.Replace("[ImageSize]", mGalleryConfig.FixedHeight.ToString & " x " & mGalleryConfig.FixedWidth.ToString)
      sb.Replace("[ThumbSize]", mGalleryConfig.MaximumThumbHeight.ToString & " x " & mGalleryConfig.MaximumThumbWidth.ToString)

      Return sb.ToString
    End Function

    Private Sub BindChildItems()

      grdContent.DataSource = mRequest.ImageList
      grdContent.PageSize = (mRequest.ImageList.Count + 1)
      grdContent.DataBind()

    End Sub

    Private Sub grdContent_ItemCommand(ByVal source As Object, ByVal e As System.Web.UI.WebControls.DataGridCommandEventArgs) Handles grdContent.ItemCommand

      Dim itemIndex As Integer = e.Item.ItemIndex
      Dim file As GalleryMaintenanceFile = CType(mRequest.ImageList.Item(itemIndex), GalleryMaintenanceFile)

      Select Case (CType(e.CommandSource, ImageButton).CommandName)
        Case "Delete"
          file.DeleteAll()
        Case "RebuildThumb"
          file.RebuildThumbnail()
        Case "CopySourceToFile"
          file.CreateFileFromSource()
        Case "CopyFileToSource"
          file.CreateFileFromSource()
        Case Else
          Throw New ArgumentException("Invalid maintenance command", "CommandName")
      End Select

      mRequest.Folder.Populate(True)
      BindData()

    End Sub

    Private Sub cmdReturn_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cmdReturn.Click
      Config.ResetGalleryConfig(ModuleId)
      Goback()
    End Sub

    Private Sub Goback()
      Dim url As String = GetURL(Page.Request.ServerVariables("URL"), Page, "", "ctl=&selectall=&mid=&currentitem=&media=")
      Response.Redirect(url)
    End Sub

    Private Function GetSelectedFiles() As List(Of GalleryMaintenanceFile)
      Dim i As Integer
      Dim selList As New List(Of GalleryMaintenanceFile)

      For i = 0 To grdContent.Items.Count - 1
        Dim myListItem As DataGridItem = grdContent.Items(i)
        Dim myCheck As CheckBox = DirectCast(myListItem.FindControl("chkSelect"), CheckBox)
        If myCheck.Checked Then selList.Add(mRequest.ImageList(i))
      Next
      Return selList

    End Function

    Private Sub RefreshAlbum()

      mRequest.Folder.Clear() ' refresh gallery folder first ready for repopulate
      mRequest.Populate()
      BindData()

    End Sub

    Private Sub btnCopySource_Click(ByVal sender As Object, ByVal e As System.Web.UI.ImageClickEventArgs)
      For Each file As GalleryMaintenanceFile In GetSelectedFiles()
        file.CreateFileFromSource()
      Next

      RefreshAlbum()

    End Sub

    Private Sub btnCopyFile_Click(ByVal sender As Object, ByVal e As System.Web.UI.ImageClickEventArgs)
      For Each file As GalleryMaintenanceFile In GetSelectedFiles()
        file.CreateSourceFromFile()
      Next

      RefreshAlbum()

    End Sub

    Private Sub btnCreateThumb_Click(ByVal sender As Object, ByVal e As System.Web.UI.ImageClickEventArgs)
      For Each file As GalleryMaintenanceFile In GetSelectedFiles()
        file.RebuildThumbnail()
      Next

      RefreshAlbum()

    End Sub

    Private Sub ClearCache1_Click(ByVal sender As Object, ByVal e As System.Web.UI.ImageClickEventArgs) Handles ClearCache1.Click
      Config.ResetGalleryConfig(ModuleId)
    End Sub

    Private Overloads Sub btnSyncAll_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSyncAll.Click
      For Each file As GalleryMaintenanceFile In GetSelectedFiles()
        file.Synchronize()
      Next

      RefreshAlbum()
    End Sub

    Private Overloads Sub btnDeleteAll_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnDeleteAll.Click
      Dim ps As PortalSettings = PortalController.GetCurrentPortalSettings()
      For Each gmFile As GalleryMaintenanceFile In GetSelectedFiles()
        Try
          If File.Exists(gmFile.AlbumPath) Then mFolder.DeleteFile(gmFile.AlbumPath, ps, True)
          If gmFile.Type = Config.ItemType.Image Then
            If File.Exists(gmFile.ThumbPath) Then mFolder.DeleteFile(gmFile.ThumbPath, ps, True)
            If File.Exists(gmFile.SourcePath) Then mFolder.DeleteFile(gmFile.SourcePath, ps, True)

            'Reset folder thumbnail to "folder.gif" if its current
            'thumbnail is being deleted. Will then get set to next image thumbnail during populate if there is one.
            If mFolder.Thumbnail = gmFile.Name Then mFolder.ResetThumbnail()
          End If

          GalleryXML.DeleteMetaData(mFolder.Path, gmFile.Name)
        Catch

        End Try
      Next

      RefreshAlbum()
    End Sub

  End Class

End Namespace

File: Maintenance.ascx.cs
using DotNetNuke;
using DotNetNuke.Common;
using DotNetNuke.Common.Utilities;
using DotNetNuke.Entities;
using DotNetNuke.Framework;
using DotNetNuke.Security;
using DotNetNuke.Services;
using DotNetNuke.Services.Exceptions;
using DotNetNuke.Services.Localization;
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics;
using System.Collections.Specialized;
using System.Configuration;
using System.Text;
using System.Text.RegularExpressions;
using System.Web;
using System.Web.Caching;
using System.Web.SessionState;
using System.Web.Security;
using System.Web.Profile;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Web.UI.WebControls.WebParts;
using System.Web.UI.HtmlControls;
//
// DotNetNuke® - http://www.dotnetnuke.com
// Copyright (c) 2002-2011 by DotNetNuke Corp. 

//
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
// documentation files (the "Software"), to deal in the Software without restriction, including without limitation 
// the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and 
// to permit persons to whom the Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all copies or substantial portions 
// of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED 
// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL 
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
// CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
// DEALINGS IN THE SOFTWARE.
//


using System.IO;
using DotNetNuke.Entities.Portals;
using DotNetNuke.Entities.Modules;
using DotNetNuke.Entities.Modules.Actions;
using static DotNetNuke.Common.Globals;
using DotNetNuke.Entities.Users;
using static DotNetNuke.Modules.Gallery.Utils;
using static DotNetNuke.Modules.Gallery.Config;
using DotNetNuke.Modules.Gallery.WebControls;

namespace DotNetNuke.Modules.Gallery
{

	public abstract partial class Maintenance : DotNetNuke.Entities.Modules.PortalModuleBase, DotNetNuke.Entities.Modules.IActionable
	{

		public string lnkViewText = "";
		#region "Private Members"
		// Obtain PortalSettings from Current Context   
		PortalSettings _portalSettings = PortalController.GetCurrentPortalSettings();
		private GalleryMaintenanceRequest mRequest;
		private DotNetNuke.Modules.Gallery.Config mGalleryConfig;
		private GalleryFolder mFolder;
		private Gallery.Authorization mGalleryAuthorization;
		private bool mSelectAll;

		private string mReturnCtl;
		//For Localization of FileInfo text
		private string mLocalizedAlbumText;
		private string mLocalizedSourceText;
		private string mLocalizedThumbText;
		private string mLocalizedMissingText;
		private string mLocalizedPresentText;
		private string mLocalizedRebuildThumbText;
		private string mLocalizedCopySourceToFileText;
		private string mLocalizedCopyFileToSourceText;
		private string mYellowWarningImageUrl;

		private string mGreenOKImageUrl;
		#endregion

		#region " Web Form Designer Generated Code "

		//This call is required by the Web Form Designer.
		[System.Diagnostics.DebuggerStepThrough()]

		private void InitializeComponent()
		{
		}

		private void Page_Init(System.Object sender, System.EventArgs e)
		{
			//CODEGEN: This method call is required by the Web Form Designer
			//Do not modify it using the code editor.
			InitializeComponent();

			if (mGalleryConfig == null) {
				mGalleryConfig = Config.GetGalleryConfig(ModuleId);
			}
			//mImageFolder = "../" & mGalleryConfig.ImageFolder.Substring(mGalleryConfig.ImageFolder.IndexOf("/DesktopModules"))

		}

		#endregion

		#region "Optional Interfaces"

		public ModuleActionCollection ModuleActions {
			get {
				ModuleActionCollection Actions = new ModuleActionCollection();
				Actions.Add(GetNextActionID(), Localization.GetString("Configuration.Action", LocalResourceFile), ModuleActionType.ModuleSettings, "", "edit.gif", EditUrl(ControlKey: "configuration"), "", false, SecurityAccessLevel.Admin, true,
				false);
				Actions.Add(GetNextActionID(), Localization.GetString("GalleryHome.Action", LocalResourceFile), ModuleActionType.ContentOptions, "", "icon_moduledefinitions_16px.gif", Common.Globals.NavigateURL(), false, SecurityAccessLevel.Edit, true, false);
				return Actions;
			}
		}

		#endregion

		public string ReturnCtl {
			get { return "Maintenance;"; }
		}

		public DotNetNuke.Modules.Gallery.Config GalleryConfig {
			get { return mGalleryConfig; }
		}


		private void Page_Load(System.Object sender, System.EventArgs e)
		{
			mRequest = new GalleryMaintenanceRequest(ModuleId);
			mGalleryConfig = mRequest.GalleryConfig;
			mFolder = mRequest.Folder;

			mGalleryAuthorization = new Authorization(ModuleId);
			if (!mGalleryAuthorization.HasItemEditPermission(mFolder)) {
				Response.Redirect(Common.Globals.AccessDeniedURL(Localization.GetString("Insufficient_Maintenance_Permissions", mGalleryConfig.SharedResourceFile)));
			}

			// Load the styles
			((CDefault)Page).AddStyleSheet(Common.Globals.CreateValidID(GalleryConfig.Css), GalleryConfig.Css);

			lnkViewText = Localization.GetString("lnkView.Text", LocalResourceFile);
			mLocalizedAlbumText = Localization.GetString("Album.Header", LocalResourceFile);
			mLocalizedSourceText = Localization.GetString("Source.Header", LocalResourceFile);
			mLocalizedThumbText = Localization.GetString("Thumb.Header", LocalResourceFile);
			mLocalizedMissingText = Localization.GetString("Missing", LocalResourceFile);
			mLocalizedPresentText = Localization.GetString("Present", LocalResourceFile);
			mLocalizedRebuildThumbText = Localization.GetString("RebuildThumb", LocalResourceFile);

			mYellowWarningImageUrl = "~/DesktopModules/Gallery/Images/yellow-warning.gif";
			mGreenOKImageUrl = "~/DesktopModules/Gallery/Images/green-ok.gif";

			//William Severance - Added condition Not Page.IsPostBack to fix loss of
			//datagrid header localization following postback. LocalizeDataGrid must be called
			//before datagrid is data bound.

			if (!Page.IsPostBack)
				Localization.LocalizeDataGrid(ref grdContent, LocalResourceFile);

			// Load gallery menu, this method allow call request only once
			GalleryMenu galleryMenu = (GalleryMenu)LoadControl("Controls/ControlGalleryMenu.ascx");
			var _with1 = galleryMenu;
			_with1.ModuleID = ModuleId;
			_with1.GalleryRequest = (BaseRequest)mRequest;
			celGalleryMenu.Controls.Add(galleryMenu);

			// Load gallery breadcrumbs, this method allow call request only once
			BreadCrumbs galleryBreadCrumbs = (BreadCrumbs)LoadControl("Controls/ControlBreadCrumbs.ascx");
			var _with2 = galleryBreadCrumbs;
			_with2.ModuleID = ModuleId;
			_with2.GalleryRequest = (BaseRequest)mRequest;
			celBreadcrumbs.Controls.Add(galleryBreadCrumbs);

			if (!Page.IsPostBack && mGalleryConfig.IsValidPath) {
				if (!mFolder.IsPopulated) {
					Response.Redirect(ApplicationURL());
				}

				BindData();

				btnDeleteAll.Attributes.Add("onClick", "javascript: return confirm('Are you sure you wish to delete selected items?')");

			}

		}


		private void BindData()
		{
			var _with3 = mFolder;
			txtPath.Text = mFolder.URL;
			txtName.Text = _with3.Name;

			lblAlbumInfo.Text = AlbumInfo();
			BindChildItems();

		}

		protected string BrowserURL(object DataItem)
		{
			GalleryMaintenanceFile item = (GalleryMaintenanceFile)DataItem;
			if (mGalleryConfig.AllowPopup) {
				return item.URL;
			} else {
				System.Collections.Generic.List<string> @params = new System.Collections.Generic.List<string>();
				@params.Add("returnctl=" + ReturnCtl);
				if ((Request.QueryString["currentstrip"] != null))
					@params.Add("currentstrip=" + Request.QueryString["currentstrip"]);
				return Utils.AppendURLParameters(item.URL, @params.ToArray());
			}
		}

		protected string FileInfo(object DataItem)
		{
			GalleryMaintenanceFile file = (GalleryMaintenanceFile)DataItem;
			System.Text.StringBuilder sb = new System.Text.StringBuilder();

			if (!file.SourceExists)
				sb.Append(mLocalizedSourceText);

			if (!file.ThumbExists) {
				if (sb.Length > 0)
					sb.Append(", ");
				sb.Append(mLocalizedThumbText);
			}

			if (!file.FileExists) {
				if (sb.Length > 0)
					sb.Append(", ");
				sb.Append(mLocalizedAlbumText);
			}

			if (sb.Length > 0) {
				sb.Append(" ");
				sb.Append(mLocalizedMissingText);
			}

			return sb.ToString();
		}

		protected string StatusImage(object DataItem, string FileType)
		{
			GalleryMaintenanceFile file = (GalleryMaintenanceFile)DataItem;
			bool exists = false;
			switch (FileType) {
				case "Thumb":
					exists = file.ThumbExists;
					break;
				case "Source":
					exists = file.SourceExists;
					break;
				case "File":
					exists = file.FileExists;
					break;
				default:
					throw new ArgumentException("Must be 'Thumb', Source' or 'File'", "FileType");
			}
			if (exists) {
				return mGreenOKImageUrl;
			} else {
				return mYellowWarningImageUrl;
			}
		}

		protected string Tooltip(object DataItem, string FileType)
		{
			GalleryMaintenanceFile file = (GalleryMaintenanceFile)DataItem;
			string localizedTooltip = mLocalizedPresentText;

			switch (FileType) {
				case "Thumb":
					if (!file.ThumbExists)
						localizedTooltip = mLocalizedRebuildThumbText;
					break;
				case "Source":
					if (!file.SourceExists)
						localizedTooltip = mLocalizedCopyFileToSourceText;
					break;
				case "File":
					if (!file.FileExists)
						localizedTooltip = mLocalizedCopySourceToFileText;
					break;
				default:
					throw new ArgumentException("Must be 'Thumb', Source' or 'File'", "FileType");
			}
			return localizedTooltip;
		}

		protected string AlbumInfo()
		{
			int itemCount = mRequest.ImageList.Count;
			int sourceCount = 0;
			int albumCount = 0;
			int thumbCount = 0;
			System.Text.StringBuilder sb = new System.Text.StringBuilder();

			foreach (GalleryMaintenanceFile file in mRequest.ImageList) {
				if (file.SourceExists)
					sourceCount += 1;
				if (file.FileExists)
					albumCount += 1;
				if (file.ThumbExists)
					thumbCount += 1;
			}

			sb.Append(Localization.GetString("AlbumInfo", LocalResourceFile));
			sb.Replace("[ItemCount]", itemCount.ToString());
			sb.Replace("[AlbumName]", mRequest.Folder.Name);
			sb.Replace("[ItemNoSource]", (itemCount - sourceCount).ToString());
			sb.Replace("[ItemNoAlbum]", (itemCount - albumCount).ToString());
			sb.Replace("[ItemNoThumb]", (itemCount - thumbCount).ToString());
			sb.Replace("[ImageSize]", mGalleryConfig.FixedHeight.ToString() + " x " + mGalleryConfig.FixedWidth.ToString());
			sb.Replace("[ThumbSize]", mGalleryConfig.MaximumThumbHeight.ToString() + " x " + mGalleryConfig.MaximumThumbWidth.ToString());

			return sb.ToString();
		}


		private void BindChildItems()
		{
			grdContent.DataSource = mRequest.ImageList;
			grdContent.PageSize = (mRequest.ImageList.Count + 1);
			grdContent.DataBind();

		}


		private void grdContent_ItemCommand(object source, System.Web.UI.WebControls.DataGridCommandEventArgs e)
		{
			int itemIndex = e.Item.ItemIndex;
			GalleryMaintenanceFile file = (GalleryMaintenanceFile)mRequest.ImageList[itemIndex];

			switch ((((ImageButton)e.CommandSource).CommandName)) {
				case "Delete":
					file.DeleteAll();
					break;
				case "RebuildThumb":
					file.RebuildThumbnail();
					break;
				case "CopySourceToFile":
					file.CreateFileFromSource();
					break;
				case "CopyFileToSource":
					file.CreateFileFromSource();
					break;
				default:
					throw new ArgumentException("Invalid maintenance command", "CommandName");
			}

			mRequest.Folder.Populate(true);
			BindData();

		}

		private void cmdReturn_Click(System.Object sender, System.EventArgs e)
		{
			Config.ResetGalleryConfig(ModuleId);
			Goback();
		}

		private void Goback()
		{
			string url = Utils.GetURL(Page.Request.ServerVariables["URL"], Page, "", "ctl=&selectall=&mid=&currentitem=&media=");
			Response.Redirect(url);
		}

		private List<GalleryMaintenanceFile> GetSelectedFiles()
		{
			int i = 0;
			List<GalleryMaintenanceFile> selList = new List<GalleryMaintenanceFile>();

			for (i = 0; i <= grdContent.Items.Count - 1; i++) {
				DataGridItem myListItem = grdContent.Items[i];
				CheckBox myCheck = (CheckBox)myListItem.FindControl("chkSelect");
				if (myCheck.Checked)
					selList.Add(mRequest.ImageList[i]);
			}
			return selList;

		}


		private void RefreshAlbum()
		{
			mRequest.Folder.Clear();
			// refresh gallery folder first ready for repopulate
			mRequest.Populate();
			BindData();

		}

		private void btnCopySource_Click(object sender, System.Web.UI.ImageClickEventArgs e)
		{
			foreach (GalleryMaintenanceFile file in GetSelectedFiles()) {
				file.CreateFileFromSource();
			}

			RefreshAlbum();

		}

		private void btnCopyFile_Click(object sender, System.Web.UI.ImageClickEventArgs e)
		{
			foreach (GalleryMaintenanceFile file in GetSelectedFiles()) {
				file.CreateSourceFromFile();
			}

			RefreshAlbum();

		}

		private void btnCreateThumb_Click(object sender, System.Web.UI.ImageClickEventArgs e)
		{
			foreach (GalleryMaintenanceFile file in GetSelectedFiles()) {
				file.RebuildThumbnail();
			}

			RefreshAlbum();

		}

		private void ClearCache1_Click(object sender, System.Web.UI.ImageClickEventArgs e)
		{
			Config.ResetGalleryConfig(ModuleId);
		}

		private void btnSyncAll_Click(System.Object sender, System.EventArgs e)
		{
			foreach (GalleryMaintenanceFile file in GetSelectedFiles()) {
				file.Synchronize();
			}

			RefreshAlbum();
		}

		private void btnDeleteAll_Click(System.Object sender, System.EventArgs e)
		{
			PortalSettings ps = PortalController.GetCurrentPortalSettings();
			foreach (GalleryMaintenanceFile gmFile in GetSelectedFiles()) {
				try {
					if (File.Exists(gmFile.AlbumPath))
						mFolder.DeleteFile(gmFile.AlbumPath, ps, true);
					if (gmFile.Type == Config.ItemType.Image) {
						if (File.Exists(gmFile.ThumbPath))
							mFolder.DeleteFile(gmFile.ThumbPath, ps, true);
						if (File.Exists(gmFile.SourcePath))
							mFolder.DeleteFile(gmFile.SourcePath, ps, true);

						//Reset folder thumbnail to "folder.gif" if its current
						//thumbnail is being deleted. Will then get set to next image thumbnail during populate if there is one.
						if (mFolder.Thumbnail == gmFile.Name)
							mFolder.ResetThumbnail();
					}

					GalleryXML.DeleteMetaData(mFolder.Path, gmFile.Name);

				} catch {
				}
			}

			RefreshAlbum();
		}
		public Maintenance()
		{
			Load += Page_Load;
			Init += Page_Init;
		}

	}

}

